### ISSUE TITLE
Can't set the resolution of a tikz plot with engine.opts

### ISSUE THREAD

**Issue by ferenci-tamas** [yihui/knitr#2114]
Here is a minimal reproducible example (that is supposed to work from [version 1.9](https://github.com/yihui/knitr/blob/master/NEWS.md)):

```
    ```{tikz,engine.opts = list(convert.opts = '-density 300')}
    \begin{tikzpicture}
    \node [] (b) {c};
    \end{tikzpicture}
    ```
```

This results in the following error message:

```
Quitting from lines 11-14 (temp.Rmd) 
Error in magick_image_format(image, toupper(format), type, colorspace,  : 
  Invalid ImageType value: -density 300
Calls: <Anonymous> ... assert_image -> do.call -> <Anonymous> -> magick_image_format
Execution halted
```

Session information:

```
R version 4.1.2 (2021-11-01)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19044), RStudio 2021.9.0.351

Locale:
  LC_COLLATE=Hungarian_Hungary.1250  LC_CTYPE=Hungarian_Hungary.1250    LC_MONETARY=Hungarian_Hungary.1250 LC_NUMERIC=C                      
  LC_TIME=Hungarian_Hungary.1250    

Package version:
  evaluate_0.15   glue_1.6.2      graphics_4.1.2  grDevices_4.1.2 highr_0.9       knitr_1.37      magrittr_2.0.2  methods_4.1.2   stats_4.1.2    
  stringi_1.7.6   stringr_1.4.0   tools_4.1.2     utils_4.1.2     xfun_0.30       yaml_2.3.5     
```


<!--
Please keep the below portion in your issue. Your issue will be closed if any of the boxes is not checked (i.e., replace `[ ]` by `[x]`). In certain (rare) cases, you may be exempted if you give a brief explanation (e.g., you are only making a suggestion for improvement). Thanks!
-->

---

By filing an issue to this repo, I promise that

- [x] I have fully read the issue guide at https://yihui.org/issue/.
- [x] I have provided the necessary information about my issue.
    - If I'm asking a question, I have already asked it on Stack Overflow or RStudio Community, waited for at least 24 hours, and included a link to my question there.
    - If I'm filing a bug report, I have included a minimal, self-contained, and reproducible example, and have also included `xfun::session_info('knitr')`. I have upgraded all my packages to their latest versions (e.g., R, RStudio, and R packages), and also tried the development version: `remotes::install_github('yihui/knitr')`.
    - If I have posted the same issue elsewhere, I have also mentioned it in this issue.
- [x] I have learned the Github Markdown syntax, and formatted my issue correctly.

I understand that my issue may be closed if I don't fulfill my promises.

**Comment by cderv** [comment]
Since version 1.22, tikz engine is now using **magick** package to do the conversion. This means that what is explain in the NEWS file for 1.9 is no more valid. 

`magick::image_convert()` is used to do the conversion and `options$engine.opts$convert.opts` is used to passed other argument to the function.  That is why you get this error as `-density 300` is passed to `type` in `magick::image_convert()` and this is not ok. 

I don't think we have any documentation about this so I understand this is not known. Maybe we should add something 🤔

**Comment by tamas-ferenci** [comment]
Thanks @cderv ! Yes, a documentation on how to set resolution with the current version would be nice.

**Comment by cderv** [comment]
> how to set resolution with the current version would be nice.

What do you want to change exactly regarding tikz default output ? `-density 300` is saying : I want 300dpi for the resulting png file ? Just trying to correctly understand so that I can see if this is possible. 

If you can show me how you would do it with **magick** R package, I could surely _translate_ in **knitr** options

**Comment by tamas-ferenci** [comment]
Yes, I thought of having a png with 300 dpi resolution. Unfortunately, I don't know the magick package either...

**Comment by cderv** [comment]
> Unfortunately, I don't know the magick package either...

I think we would need to understand how to do this using **magick** 📦 and then we could see how this could be done in a Rmd document with _tikz_ engine. 

You could ask probably on this forum https://discuss.ropensci.org/tags/c/packages/7/magick or in github repo maybe (I would give preference to the forum). 

What is done in **knitr** is 
* read the pdf
* convert to png
* write the png to file

I don't know exactly if a DPI conversion is made, we are currently not modifying anything in this process. PDF is read with 300 density, so I guess it should be kept when writing png. 

I'll leave that open but we won't be able to help more on this for now.

**Comment by tamas-ferenci** [comment]
Thank you @cderv !

**Comment by justanothergithubber** [comment]
Hello @cderv, [as I understand it](https://github.com/yihui/knitr/blob/9860f1cfdf220b253bcb60bc2590dfe90e4eb683/R/engine.R#L386) indeed `magick::image_read_pdf` [(src)](https://github.com/ropensci/magick/blob/52df93df7779e84b29d34971b33b1c0986bc2b5d/R/edit.R#L133) is called with its default `density` argument. If possible, could some density argument be added? It would likely need to propagate all the way to the outer `magick::image_write` call as well, so perhaps something like
```
eng_tikz = function(options) {
...
    if (ext != 'pdf') magick::image_write(do.call(magick::image_convert, c(
      list(magick::image_read_pdf(fig, density=options$density), ext), options$engine.opts$convert.opts
    )), fig2, density=options$density)
...
}
```
This would greatly help with outputs and make them a lot nicer. If you have a nicer implementation of this, that would be great too.

**Comment by cderv** [comment]
Hi @justanothergithubber ,

Can you share the R code you would like to be executed when producing an image ? I could then try to see clearer what is missing in **knitr**. 

Basically we are converting PDF to another format (like PNG) using imagemagick through the magick package
1. `magick::image_read_pdf()` will read the PDF by producing images of 300DPI
2. `magick::image_convert()` will convert to PNG, and **knitr** allows customization of the convertion
3. `magick::image_write()` will write to file using default density

If I understand correctly, you are saying that DPI can't be changed at step 2 above, and needs to be made configurable using chunk option at step 1 or 3. Is that right ? 

Thanks

**Comment by justanothergithubber** [comment]
Hello again,

I am currently saying that the DPI cannot be changed at step 1, `magick::image_read_pdf()`, which affects output. I must admit I do not know how much an argument to `magick::image_write()` will matter, although I guess to people who *really* want to customise their output, it could be a feature they want.

Further below I have made a toy tikz example and I have generated images at read densities of 300, 600 and 1200 respectively. I would prefer being able to use 1200 as I typically scale the entire image to the width if possible (though I probably did not try images from 600 to 1200), although I understand the image is correspondingly bigger, in this case about 4x as big. The engine is `tikz`
```{tikz}
\begin{tikzpicture}
\usetikzlibrary{positioning}
\node (A) at (0, 0) {A};
\node (B) at (0, 1) {B};
\node (C) at (1, 0) {C};
\node (D) at (1, 1) {D};
\end{tikzpicture}
````
![300](https://user-images.githubusercontent.com/3600019/169849576-81666644-d5ec-4516-84de-b7903916ebf8.png)
![600](https://user-images.githubusercontent.com/3600019/169849574-278edd26-396c-4a37-955e-4a694280538e.png)
![1200](https://user-images.githubusercontent.com/3600019/169849565-a2be0d3c-ea0a-4b11-829d-31a31b943ba3.png)

**Comment by cderv** [comment]
> I am currently saying that the DPI cannot be changed at step 1, magick::image_read_pdf(), which affects output.

Great, that is what I wanted to confirm. I'll see the best way to add this configuration available at chunk level.

**Comment by oliviermeslin** [comment]
Hi @cderv , I am not sure whether you have already modified anything to solve this issue so what I write below might be irrelevant.

My impression is that it is quite difficult to modify the quality of the pdf-to-png conversion with `magick ::image_convert` because this function does not have any `density` option. However, the `pdftools::pdf_convert` has a density option, so maybe changing the underlying tool used for conversion might be an option? Or you might introduce a new chunk option to specify the _package_ used for conversion?

**Comment by yihui** [comment]
Fixed. The `tikz` engine now supports an `engine.opts$density` element to set the rasterization resolution (in DPI), e.g.:

````
```{tikz, engine.opts = list(density = 600)}
\begin{tikzpicture}
\node (a) {A};
\end{tikzpicture}
```
````

The old `convert.opts = '-density 300'` never worked after the engine switched from command-line ImageMagick to the **magick** package, so this replaces it with a dedicated option (defaulting to 300 DPI, the previous fixed value). Thanks for the report, @ferenci-tamas, and @justanothergithubber for the diagnosis!
